File: java_src/main/java/com/example/lab3/MainActivity.java
package com.example.lab3;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {
    private static final int SCISSORS = 0;
    private static final int STONE = 1;
    private static final int PAPER = 2;

    private EditText ed_name;
    private TextView tv_text, tv_name, tv_winner, tv_mmora, tv_cmora;
    private RadioButton btn_scissor, btn_stone, btn_paper;
    private Button btn_mora;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initializeViews();

        btn_mora.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                playGame();
            }
        });
    }
    private void initializeViews() {
        ed_name = findViewById(R.id.ed_name);
        tv_text = findViewById(R.id.tv_text);
        tv_name = findViewById(R.id.tv_name);
        tv_winner = findViewById(R.id.tv_winner);
        tv_mmora = findViewById(R.id.tv_mmora);
        tv_cmora = findViewById(R.id.tv_cmora);
        btn_scissor = findViewById(R.id.btn_scissor);
        btn_stone = findViewById(R.id.btn_stone);
        btn_paper = findViewById(R.id.btn_paper);
        btn_mora = findViewById(R.id.btn_mora);
    }
    private void playGame() {
        String playerName = ed_name.getText().toString();

        if (TextUtils.isEmpty(playerName)) {
            tv_text.setText("請輸入玩家姓名");
            return;
        }

        int playerMove = getPlayerMove();
        int computerMove = (int) (Math.random() * 3);

        updateMovesUI(playerName, playerMove, computerMove);

        determineWinner(playerName, playerMove, computerMove);
    }
    private int getPlayerMove() {
        if (btn_scissor.isChecked()) {
            return SCISSORS;
        } else if (btn_stone.isChecked()) {
            return STONE;
        } else {
            return PAPER;
        }
    }
    private String getMoveString(int move) {
        switch (move) {
            case SCISSORS:
                return "剪刀";
            case STONE:
                return "石頭";
            case PAPER:
                return "布";
            default:
                return "";
        }
    }
    private void updateMovesUI(String playerName, int playerMove, int computerMove) {
        tv_name.setText(String.format("名字\n%s", playerName));
        tv_mmora.setText(String.format("我方出拳\n%s", getMoveString(playerMove)));
        tv_cmora.setText(String.format("電腦出拳\n%s", getMoveString(computerMove)));
    }
    private void determineWinner(String playerName, int playerMove, int computerMove) {
        if (playerMove == computerMove) {
            tv_winner.setText("勝利者\n平手");
            tv_text.setText("平局，請再試一次！");
        } else if ((playerMove == PAPER && computerMove == STONE) ||
                (playerMove == STONE && computerMove == SCISSORS) ||
                (playerMove == SCISSORS && computerMove == PAPER)) {
            tv_winner.setText("勝利者\n" + playerName);
            tv_text.setText("恭喜您獲勝了！！！");
        } else {
            tv_winner.setText("勝利者\n電腦");
            tv_text.setText("可惜，電腦獲勝了！");
        }
    }
}

File: Kotlin_src/main/java/com/example/lab3/MainActivity.kt
package com.example.lab3

import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.RadioGroup
import android.widget.TextView
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat

class MainActivity : AppCompatActivity() {
    private lateinit var edName: EditText
    private lateinit var tvText: TextView
    private lateinit var radioGroup: RadioGroup
    private lateinit var btnMora: Button
    private lateinit var tvName: TextView
    private lateinit var tvWinner: TextView
    private lateinit var tvMyMora: TextView
    private lateinit var tvTargetMora: TextView

    // 在 companion object 中定義常數，取代 magic numbers
    private companion object {
        const val SCISSORS = 0
        const val STONE = 1
        const val PAPER = 2
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        initializeViews()

        btnMora.setOnClickListener {
            playGame()
        }
    }

    private fun initializeViews() {
        edName = findViewById(R.id.edName)
        tvText = findViewById(R.id.tvText)
        radioGroup = findViewById(R.id.radioGroup)
        btnMora = findViewById(R.id.btnMora)
        tvName = findViewById(R.id.tvName)
        tvWinner = findViewById(R.id.tvWinner)
        tvMyMora = findViewById(R.id.tvMyMora)
        tvTargetMora = findViewById(R.id.tvTargetMora)
    }

    private fun playGame() {
        val playerName = edName.text.toString()

        if (playerName.isBlank()) {
            tvText.text = "請輸入玩家姓名"
            return
        }

        val playerMove = getPlayerMove()
        val computerMove = (SCISSORS..PAPER).random() // 語意更清晰的隨機數產生

        updateMovesUI(playerName, playerMove, computerMove)

        determineWinner(playerName, playerMove, computerMove)
    }
    private fun getPlayerMove(): Int {
        return when (radioGroup.checkedRadioButtonId) {
            R.id.btnScissor -> SCISSORS
            R.id.btnStone -> STONE
            else -> PAPER
        }
    }
    private fun getMoveString(move: Int): String {
        return when (move) {
            SCISSORS -> "剪刀"
            STONE -> "石頭"
            else -> "布"
        }
    }
    private fun updateMovesUI(playerName: String, playerMove: Int, computerMove: Int) {
        tvName.text = "名字\n$playerName"
        tvMyMora.text = "我方出拳\n${getMoveString(playerMove)}"
        tvTargetMora.text = "電腦出拳\n${getMoveString(computerMove)}"
    }
    private fun determineWinner(playerName: String, playerMove: Int, computerMove: Int) {
        when {
            playerMove == computerMove -> {
                tvWinner.text = "勝利者\n平手"
                tvText.text = "平局，請再試一次！"
            }
            (playerMove == PAPER && computerMove == STONE) ||
                    (playerMove == STONE && computerMove == SCISSORS) ||
                    (playerMove == SCISSORS && computerMove == PAPER) -> {
                tvWinner.text = "勝利者\n$playerName"
                tvText.text = "恭喜你獲勝了！！！"
            }
            else -> {
                tvWinner.text = "勝利者\n電腦"
                tvText.text = "可惜，電腦獲勝了！"
            }
        }
    }
}